Quantile methods

hkershaw-brown opened the pull request:
## Description:
<!--- Describe your changes -->
Adds a Quantile-Conserving Ensemble Filtering Framework to DART
Publications: [QCEFF part1](http://n2t.net/ark:/85065/d7mk6hm4) [QCEFF part 2](http://n2t.net/ark:/85065/d7nv9pbt).

The default QCEFF options are EAKF, normal distribution (no bounds).

**This is a breaking change and will be v11 if DART**
Namelist item removed from filter_nml
   * filter_kind - now a per qty option
   
Two new required namelists (added to input.nml files):
* &probit_transform_nml
* &algorithm_info_nml

assim_tools_mod
* sort_obs_inc namelist option applied to ENKF only
* ``spread_restoration`` is not supported in this version

algorithm_info_mod
* runtime QCEFF options read from .csv or .txt file

New probability distribution  modules:
* beta_distribution_mod
* bnrh_distribution_mod (bounded normal rank histogram)
* gamma_distribution_mod
* normal_distribution_mod

* probit_transform_mod
* distribution_params_mod

Update to lorenz_96_tracer_advection:
* positive_tracer
* default mean velocity = 0)
* perturbation
* create_obs_sequence input file

1d forward operator updates:
* For non-integer powers, fix up values for negative bases

Docs:
* index.rst: Nonlinear and Non-Gaussian Data Assimilation Capabilities in DART
* QCEFF instructions: Quantile-Conserving Ensemble Filter Framework
* Example to work through: QCEFF: Examples with the Lorenz 96 Tracer Model

### Fixes issue
<!--- link to github issue(s) -->

review notes #458 
Note this pull request includes the fix for #416

Todo:
- [ ] continuing existing experiments #477
- [ ] Remove models/lorenz_96/work/README (readme file for papers)

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?
<!-- Put an `x` in all the boxes that apply: -->
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.

### Tests
Please describe any tests you ran to verify your changes.
bitwise lorenz_96 with main for default qcf options

### Timing:
* main vs. qceff default  [Derecho](https://docs.google.com/spreadsheets/d/1eWfTvj4vsmqbyguaAtWL4QbM4bobHdNc_CI0-r0VtWc/edit#gid=0)
* bhrf vs main [Cheyenne](https://docs.google.com/document/d/1njqMWIFxAy13WnkDmZhnkI8aqYs6a0u4jmJmo44np2g/edit)

## Checklist for merging

- [ ] Updated changelog entry
- [ ] Documentation updated
- [ ] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

hkershaw-brown commented:
note to reviewers, documentation for this pull request is here:
https://dart-documentation--588.org.readthedocs.build/en/588/

hkershaw-brown commented:
@jlaucar can you go ahead and add whoever you want to the list of reviewers. Cheers,
Helen

hkershaw-brown commented on assimilation_code/modules/assimilation/assim_tools_mod.rst:
```diff
@@ -10,25 +10,6 @@ for both mean and spread. In addition, algorithms to do a variety of flavors of
--  9 = Particle filter (see Poterjoy 2016)
-
-We recommend using type=1, the EAKF. Note that although the algorithm is expressed in a slightly different form, the
-EAKF is identical to the EnSRF (Ensemble Square Root Filter) described by Whitaker and Hamill in 2002. Highly
-non-gaussian distributions may get better results from type=8, Rank Histogram filter.
```
Does this note on EAKF vs EnSRF need adding to the QCEFF filter option docs:
https://docs.dart.ucar.edu/en/quantile_methods/guide/qceff_probit.html#available-filter-kinds ?

hkershaw-brown replied:
discussed with jla
can pul this EnSRF comment out.

hkershaw-brown commented on assimilation_code/modules/assimilation/algorithm_info_mod.f90:
```diff
@@ -0,0 +1,620 @@
+integer, parameter :: EAKF               = 1
+integer, parameter :: ENKF               = 2
+integer, parameter :: KERNEL             = 3
+integer, parameter :: OBS_PARTICLE       = 4
```
KERNEL, OBS_PARTICLE are not selectable as filter_kinds
see line 282 case statement 
      case ('EAKF')
      case ('ENKF')
      case ('UNBOUNDED_RHF')
      case ('GAMMA_FILTER')
      case ('BOUNDED_NORMAL_RHF')

hkershaw-brown commented on assimilation_code/modules/assimilation/filter_mod.f90:
```diff
@@ -351,13 +359,13 @@ subroutine filter_main()
 call check_namelist_read(iunit, io, "filter_nml")
 
+call filter_initialize_modules_used() ! static_init_model called in here
+
```
done

hkershaw-brown commented on assimilation_code/modules/assimilation/distribution_params_mod.f90:
```diff
@@ -0,0 +1,49 @@
+
+! Provides data structure and tools to represent probability distribution families for DART
+
+use types_mod, only : r8, missing_r8
```
```suggestion
use types_mod, only : r8
```

unused missing_r8

hkershaw-brown commented on assimilation_code/modules/assimilation/filter_mod.f90:
```diff
@@ -50,7 +50,7 @@ module filter_mod
                                   all_copies_to_all_vars, allocate_single_copy, allocate_vars, &
                                   get_single_copy, put_single_copy, deallocate_single_copy,   &
-                                  print_ens_handle
+                                  print_ens_handle, get_my_vars
```
```suggestion
                                  print_ens_handle
```
unused get_my_vars

hkershaw-brown commented on assimilation_code/modules/assimilation/assim_tools_mod.f90:
```diff
@@ -71,13 +71,32 @@ module assim_tools_mod
           set_assim_tools_trace, &
           test_state_copies, &
           update_ens_from_weights
+! Test functions
+public :: test_get_state_meta_data, test_close_obs_dist
```
these test_subroutines are used only inside assim_tools_mod.f90

johnsonbk reviewed (commented):
I fixed the typos mentioned in the comments of this review in PR six twelve. Approving the six twelve merge will reconcile the typos and will add substantive edits to `index.rst`.

I added the substantive edits to index.rst because the quantile methods merge adds a lot of new information to the documentation that DART users may be unfamiliar with. It may be easier to read by following a standard practice in technical writing: using topic and stress positions in sentences. The topic position in a sentence contains information that a user already knows, or is presumed to know, while the stress position introduces new information.

After running through the four examples in `guide/qceff-examples.rst` and plotting their output with diagnostics/matlab/plot_ens_time_series.m, my conclusion is that

1. Further exposition to the QCEFF examples should be added, or
2. The examples should be modified to make the benefits of the new framework more evident

mjs2369 commented:
These comments are now incorrect and should be removed since we are once again dividing by sqrt(2 PI)

[https://github.com/NCAR/DART/blob/d5c21646ebef17865cfe779fd68894cbe6a51112/assimilation_code/modules/assimilation/assim_tools_mod.f90#L1701C1-L1702C48](https://github.com/NCAR/DART/blob/d5c21646ebef17865cfe779fd68894cbe6a51112/assimilation_code/modules/assimilation/assim_tools_mod.f90#L1701C1-L1702C48)

[https://github.com/NCAR/DART/blob/d5c21646ebef17865cfe779fd68894cbe6a51112/assimilation_code/modules/assimilation/assim_tools_mod.f90#L1714C1-L1715C48](https://github.com/NCAR/DART/blob/d5c21646ebef17865cfe779fd68894cbe6a51112/assimilation_code/modules/assimilation/assim_tools_mod.f90#L1714C1-L1715C48)

mjs2369 commented on guide/qceff-examples.rst:
```diff
@@ -0,0 +1,184 @@
+.. code-block:: text
+
+   git checkout quantile_methods
+
```
These lines above (7-15) shoud be removed or edited before release. The tools will no longer be in the alpha release stage and the user will not have to checkout the quantile_methods branch

```
The Quantile-Conserving Ensemble Filter Framework (QCEFF) tools are in the alpha release stage.
The DART development team (dart@ucar.edu) would be happy to hear about your experiences and is
anxious to build scientific collaborations using these new capabilities.


To get started, make sure that you are on the quantile_methods branch of DART: 


.. code-block:: text


   git checkout quantile_methods

```

There is another comment like this at the very beginning of guide/qceff_probit.rst

hkershaw-brown replied:
Since this branch is the live documentation, I'll remove the warnings just before merging.
I plopped them in the checklist:

![Screen Shot 2024-01-05 at 9 55 53 AM](https://github.com/NCAR/DART/assets/20047007/b6d21c83-8795-4776-85b0-8935996b71c9)

Thanks for the reminder (and future reminders so I/you/we don't miss this)

mjs2369 reviewed (commented):
The &probit_transform_nml and &algorithm_info_nml are missing from models/MITgcm_ocean/work/input.nml

mjs2369 commented on assimilation_code/modules/assimilation/probit_transform_mod.f90:
```diff
@@ -0,0 +1,733 @@
+
+!!!elseif(p%distribution_type == PARTICLE_FILTER_DISTRIBUTION) then
+   !!!call to_probit_particle(ens_size, state_ens, p, probit_ens, use_input_p, &
+       !!!bounded_below, bounded_above, lower_bound, upper_bound)
```
Are we still planning on supporting `distribution_type = PARTICLE_FILTER_DISTRIBUTION ` in the future?

hkershaw-brown replied:
remove PARTICLE_FILTER_DISTRIBUTION from list of available distributions 

https://github.com/NCAR/DART/blob/d5c21646ebef17865cfe779fd68894cbe6a51112/guide/qceff_probit.rst?plain=1#L123-L132

since it is not available in this code.

hkershaw-brown replied:
@mjs2369 I'm not sure when PARTICLE_FILTER_DISTRIBUTION will be supported, 
I believe @jlaucar has experimented/tested with this. 
For now I fixed the docs to match the code. Maybe PARTICLE_FILTER_DISTRIBUTION  goes in v11.1?

mjs2369 replied:
Yep I agree that the PARTICLE_FILTER_DISTRIBUTION can be put into a later release if we want. Good catch with removing it from this release of the documentation. 

Should we also remove it from this comment in algorithm_info_mod?
[https://github.com/NCAR/DART/blob/9e6b2068602d302aae4b5d12480b7a8dd2bade1b/assimilation_code/modules/assimilation/algorithm_info_mod.f90#L386](https://github.com/NCAR/DART/blob/9e6b2068602d302aae4b5d12480b7a8dd2bade1b/assimilation_code/modules/assimilation/algorithm_info_mod.f90#L386)

hkershaw-brown replied:
@mjs2369 it is not just the comment right? It is the code in algortihm_info_mod also:
https://github.com/NCAR/DART/blob/9e6b2068602d302aae4b5d12480b7a8dd2bade1b/assimilation_code/modules/assimilation/algorithm_info_mod.f90#L249-L250

@jlaucar do you want particle_filter_distritbution in this release?

mjs2369 replied:
Yes we should remove that as well. I feel like we the code would potentially take a while to be ready for this release and we should probably put it in with a later one

If we do as such, should we also remove the commented out code pertaining to it in probit_transform_mod? Lines 187-9 and the subroutine to_probit_particle starting on line 378

hkershaw-brown replied:
@mjs2369 chatted with Jeff about this. 
Conclusion: don't let people select PARTICLE_FILTER_DISTRIBUTION.  Leave in the commented subroutine to_probit_particle (if we were being strict about code we would remove this, but this is science).  I'll go ahead and remove `case ('PARTICLE_FILTER_DISTRIBUTION') ` algorithm_info_mod.f90.   Once again thanks for the detailed review on this. 

Cheers,
Helen

hkershaw-brown commented on guide/qceff-examples.rst:
```diff
@@ -0,0 +1,184 @@
+
+   * - example
+     - description
+     - .cvs filename 
```
fixed in #612

hkershaw-brown commented on guide/qceff-examples.rst:
```diff
@@ -0,0 +1,184 @@
+     - description
+     - .cvs filename 
+   * - Example A 
+     - boundend normal rank histogram
```
fixed in https://github.com/NCAR/DART/pull/612

hkershaw-brown commented on guide/qceff_probit.rst:
```diff
@@ -0,0 +1,152 @@
+
+To add a quantity, add a row to the table.
+For any quantity not listed in the table, the :ref:`Default values` values will be used for all 25 options. 
+You only have to add rows for quantitiies that use non-default values for any of the input options.
```
fixed in https://github.com/NCAR/DART/pull/612

hkershaw-brown commented on guide/qceff_probit.rst:
```diff
@@ -0,0 +1,152 @@
+These two headers must be present in your qceff table. 
+The :ref:`qcf trunc table` below shows the table headers, 
+and an example quantity QTY_TRACER_CONCENTRATION for the first 5 columns of the table. 
+There is a complete table with all 25 columns in `Google Sheets <https://docs.google.com/spreadsheets/d/1CRGHWc7boQt81pw2pDxEFY6WPyQeCh64OwPyoVMqijE/edit?usp=sharing>`_. You can copy and edit this table as needed.
```
@mjs2369 I changed QTY_NAME: to QTY_EXAMPLE to address Ben's comment. 
![Screen Shot 2024-01-05 at 9 21 56 AM](https://github.com/NCAR/DART/assets/20047007/07589029-e85b-4722-a1d6-084b51c07491)

If this isn't sufficient to show that this is an example of a qty change it to whatever is appropriate.

hkershaw-brown replied:
It does not matter - the header line needs to be there.

mjs2369 replied:
I think the change to QTY_EXAMPLE is good.

I think what Brett is asking is if we should make all of the example CSVs in the lorenz_96_tracer_advection/work dir consistent in their headers in cell 2A, so we don't confuse people. We should make them all match the newest qcf_table_template (v1), which uses QTY_NAME instead of QTY_TEMPLATE. So all_bnrhf_qceff_table.csv, all_eakf_qceff_table.csv, and neg_qceff_table.csv would all get edited. 

We also need to to update the format when logging the qceff_table to fit QTY_NAME:
[https://github.com/NCAR/DART/blob/9e6b2068602d302aae4b5d12480b7a8dd2bade1b/assimilation_code/modules/assimilation/algorithm_info_mod.f90#L581](https://github.com/NCAR/DART/blob/9e6b2068602d302aae4b5d12480b7a8dd2bade1b/assimilation_code/modules/assimilation/algorithm_info_mod.f90#L581)

The new formatting will be:
`write(log_msg,'(A9, 2A14, 2A12, 3(A10, 2A14, 2A12), A12, 2A14, 2A12)') header2(:)`

mjs2369 replied:
Also I believe that we should change the name of the Google Sheet template from qcf_table_template to qceff_table_template

And cell A1 should be changed to "QCEFF table version: 1"

This would require us to change the log formatting for the first header here as well:
[https://github.com/NCAR/DART/blob/9e6b2068602d302aae4b5d12480b7a8dd2bade1b/assimilation_code/modules/assimilation/algorithm_info_mod.f90#L578](https://github.com/NCAR/DART/blob/9e6b2068602d302aae4b5d12480b7a8dd2bade1b/assimilation_code/modules/assimilation/algorithm_info_mod.f90#L578)
to 
`write(log_msg, '(A6, A6, A9, A)') header1(:)`

And update the headers for all of the .csv files in lorenz_96_tracer_advection/work to "QCEFF table version: 1"

hkershaw-brown replied:
ok @mjs2369  can you go ahead and make the changes to the google sheet, contents and name, and change the example .csv files in lorenz_96_tracer.

hkershaw-brown commented:
> I fixed the typos mentioned in the comments of this review in PR six twelve. Approving the six twelve merge will reconcile the typos and will add substantive edits to `index.rst`.
> 
> I added the substantive edits to index.rst because the quantile methods merge adds a lot of new information to the documentation that DART users may be unfamiliar with. It may be easier to read by following a standard practice in technical writing: using topic and stress positions in sentences. The topic position in a sentence contains information that a user already knows, or is presumed to know, while the stress position introduces new information.
> 

 doc changes merge in from  #612 pull request.

hkershaw-brown commented:
> The &probit_transform_nml and &algorithm_info_nml are missing from models/MITgcm_ocean/work/input.nml

Add these in [8ed6c2e](https://github.com/NCAR/DART/pull/588/commits/8ed6c2e639e732071d000f317a2f748e3aa9fef3)

Add for any model/input.nml that was missing these.

```
#!/bin/bash
files=$(find . -name input.nml)

for f in $files
do
   if ! grep -q probit_transform_nml  $f; then
     echo probit not found $f
   fi

done

for f in $files
do
  if ! grep -q algorithm_info_nml  $f; then
     echo algorithm not found $f
   fi  
done
```

jlaucar commented on index.rst:
```diff
@@ -55,6 +61,56 @@ These tools are intended for use by the full range of geosciencies community:
+and the observation error. For a detailed description of the QCEFF, see
+Anderson (2022). [2]_ 
+
+While the QCEFF for computing observation increments can lead to significant improvements in 
+analysis estimates for observed variables, those improvements can be lost when using standard 
+linear regression of observation increments to update other state variables. The QCEFF also 
+implements a capability to do regression in a probit probability integral transformed space. 
+Doing the regression of observation quantile increments in the transformed space guarantees 
+that the posterior ensembles for state variables also retain the advantages of the observation space
+quantile conserving posteriors. For example, if state variables are bounded, then posterior 
+ensembles will respect the bounds. The posterior ensembles also respect other aspects of the 
+continuous prior distributions. For a detailed description of this process, see
```
Perhaps the easiest way to do this is to find a way to link to the draft to part 3 of the QCEFF paper series which contatins a number of figures comparing the methods and also finalizes the documentation of the method. I don't know how hard it is to find a place to stuff the draft paper while it is going through review, so not sure that this is the solution.

hkershaw-brown replied:
draft available & added citation to docs in 11457dec4fdeaf8f6e39e6db98c05a1fe2a6b25b

jlaucar commented on index.rst:
```diff
@@ -55,6 +61,56 @@ These tools are intended for use by the full range of geosciencies community:
+
+While the QCEFF for computing observation increments can lead to significant improvements in 
+analysis estimates for observed variables, those improvements can be lost when using standard 
+linear regression of observation increments to update other state variables. The QCEFF also 
+implements a capability to do regression in a probit probability integral transformed space. 
+Doing the regression of observation quantile increments in the transformed space guarantees 
+that the posterior ensembles for state variables also retain the advantages of the observation space
```
This is a good catch. I think the solution is to drop 'other' rather than replacing it with 'unobserved'.  All state variables, observed or unobserved, are updated by regression of the increments for the current observation.

hkershaw-brown replied:
fixed in [57652b2](https://github.com/NCAR/DART/pull/588/commits/57652b21c46fd9485862d10073c09485246fc2b1)

jlaucar commented on guide/qceff_probit.rst:
```diff
@@ -0,0 +1,153 @@
+
+* Observation error information
+
+   Used to compute sample for this observation when using perfect_model_obs
```
Another good catch. I suggest "Provides information about boundedness constraints that control the likelihood distribution associated with an observed variable.."

hkershaw-brown replied:
put Jeff's suggestion in [76202f3](https://github.com/NCAR/DART/pull/588/commits/76202f360f1067032c26c85f360136b26bae540d)

jlaucar commented on guide/qceff-examples.rst:
```diff
@@ -0,0 +1,184 @@
+
+Assimilating observations of state (wind) and tracer concentration using
+a rank histogram observation space filter and rank histogram probit transforms for
+state variable updates. This example includes adaptive inflation.
```
I think this is okay as is. All state variables are updated in this way.

jlaucar commented on guide/qceff-examples.rst:
```diff
@@ -0,0 +1,184 @@
+      ``./filter``
+
+#. Examine the output with your favorite tools. Looking at the analysis ensemble 
+   for the tracer_concentration variables with indices near the source (location 1)
+   and far downstream from the source (location 35) is interesting. Note that the
```
How about adding "Near the source, the true concentration and the ensemble estimates are all non-zero while far from the source there are times when the true concentration and many ensemble members are zero. "  

Detail on this is also provided in the part 3 paper if we can find a way to link it.

hkershaw-brown replied:
Added Jeffs "Near the source ..."
& qceff III draft available & added citation to docs in 11457dec4fdeaf8f6e39e6db98c05a1fe2a6b25b
